### PR TITLE
DT v2 + Project structure upgrades

### PR DESCRIPTION
**Pull request by shanmukhateja**
Closes #1777 and #1780 

Feel free to test this branch for issues. I have verified the functionality using `npm test`.

Also, I suggest cloning this branch in a separate directory due to differences in project structure and `devDependencies`.

### CODE REVIEW

**Comment by shanmukhateja** [comment]
@l-lin I may have broken GitHub CI flow with these changes. We simply need to update `npm` commands used in the workflow file and test it back-to-back.

If you can spare some time, please look into this. It might take me a while to check this due to some personal commitments. :-^)

**Comment by shanmukhateja** [comment]
I still need to update documentation pages and DEVELOPER.md

File: package.json [hunk 1]
```diff
@@ -24,45 +24,77 @@
...
     "Surya Teja K <shanmukhateja@gmail.com>"
   ],
-  "schematics": "./schematics/src/collection.json",
-  "main": "bundles/angular-datatables.umd.js",
-  "module": "index.js",
-  "typings": "index.d.ts",
+  "schematics": "./lib/schematics/src/collection.json",
+  "main": "src/index.js",
+  "module": "src/index.js",
+  "typings": "src/index.d.ts",
```
Comment (l-lin) [inline package.json:30]: __issue__: I think the folder is not correct as there's no `src` folder.

File: package.json [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
-    "compile": "ngc -p tsconfig-build.json",
-    "compile:tsc": "tsc -p tsconfig.json",
-    "bundles": "npm run rollup && npm run rollup:min",
-    "schematics:build": "tsc -p schematics/tsconfig.json",
-    "rollup": "rollup -c rollup.conf.js",
-    "rollup:min": "rollup -c rollup-uglify.conf.js",
+    "start": "npm run link:lib && ng serve",
+    "build:lib": "npm run clean && ng build angular-datatables --configuration=production && npm run lib:schematics:build",
+    "clean": "rimraf -f lib/**/index.{d.ts,js,js.map,metadata.json} demo/**/*.{d.ts,js,map,metadata.json} schematics/**/*.{d.ts,js,map}",
+    "link:lib": "rimraf node_modules/angular-datatables && npm run build:lib && linklocal",
```
Comment (l-lin) [inline package.json:9]: __chore__: Can you also update/delete the `demo/README.md` that still has reference on the `install-local` tool?

File: DEVELOPER.md [hunk 3]
```diff
@@ -40,11 +40,9 @@ The `npm run build` command compiles the library,
```
Comment (l-lin) [inline DEVELOPER.md:40]: __chore__: The instructions above are obsolete and need to be updated to use the new commands.

File: demo/src/app/app.component.css [hunk 4]
```diff
@@ -9,8 +9,9 @@ header .container a {
 }
```
Comment (l-lin) [inline demo/src/app/app.component.css:9]: __issue__: I don't know if there's something missing, when I ran the demo with `npm start`, the search bar is not well rendered:

image

**Review by l-lin — CHANGES_REQUESTED** [review]
__chore__: We need another change on the documentation where we need to specify that angular-datatables will only support datatables v2 starting from version X.

**Comment by shanmukhateja** [comment]
Hey @l-lin 

Glad to know you're back.

I have just closed the laptop for the night so let me check and update the PR tomorrow (IST timezone).

:)

File: demo/src/app/extensions/buttons-extension.component.ts [hunk 5]
```diff
@@ -40,7 +41,7 @@ export class ButtonsExtensionComponent implements OnInit {
           extend: 'csv',
           text: 'CSV export',
           fieldSeparator: ';',
-          exportOption: [1, 2, 3]
+          exportOptions: [1, 2, 3]
```
[the remark this answers is not in the document]
Comment (shanmukhateja) [reply demo/src/app/extensions/buttons-extension.component.ts:44]: In DT v2, the property name was changed.

File: demo/src/app/extensions/colreorder-extension.component.ts [hunk 6]
```diff
@@ -32,7 +33,6 @@ export class ColreorderExtensionComponent implements OnInit {
       // Use this attribute to enable colreorder
       colReorder: {
         order: [1, 0, 2],
-        fixedColumnsRight: 2
```
[the remark this answers is not in the document]
Comment (shanmukhateja) [reply demo/src/app/extensions/colreorder-extension.component.ts:35]: Fixed :)

File: lib/public_api.ts [hunk 7]
```diff
@@ -0,0 +1 @@
+export * from './index';
```
[the remark this answers is not in the document]
Comment (shanmukhateja) [reply lib/public_api.ts]: Fixed :)

File: package.json (hunk 1 above)
Comment (shanmukhateja) [reply package.json]: Good catch! Fixed :)

File: package.json (hunk 2 above)
Comment (shanmukhateja) [reply package.json:9]: Fixed :)

File: DEVELOPER.md (hunk 3 above)
Comment (shanmukhateja) [reply DEVELOPER.md:40]: Fixed :)

File: demo/src/app/app.component.css (hunk 4 above)
Comment (shanmukhateja) [reply demo/src/app/app.component.css:9]: This is happening due to incompatibility between materializecss and datatables.net-dt styling.

Fixed :)

**Comment by shanmukhateja** [comment]
@l-lin Please don't merge the PR just yet. I need to test `ng add` with Bootstrap + DT v2.

File: demo/src/app/extensions/responsive-extension.component.html [hunk 8]
```diff
@@ -1,6 +1,5 @@
 <ng-template #preview>
   <table datatable [dtOptions]="dtOptions" class="row-border hover"></table>
 </ng-template>
-<app-base-demo [pageTitle]="pageTitle" [mdIntro]="mdIntro" [mdInstall]="mdInstall" [mdHTML]="mdHTML" [mdTS]="mdTS"
-  [template]="preview">
+<app-base-demo [pageTitle]="pageTitle" [mdIntro]="mdIntro" [mdInstall]="mdInstall" [mdHTML]="mdHTML" [mdTS]="mdTS" mdTSV2="mdTSV2" [template]="preview">
```
Comment (l-lin) [inline demo/src/app/extensions/responsive-extension.component.html:4]:
```suggestion
<app-base-demo [pageTitle]="pageTitle" [mdIntro]="mdIntro" [mdInstall]="mdInstall" [mdHTML]="mdHTML" [mdTS]="mdTS" [mdTSV2]="mdTSV2" [template]="preview">
```

__issue__: Missing brackets `[]`.

**Review by l-lin — CHANGES_REQUESTED** [review]
> **chore**: We need another change on the documentation where we need to specify that angular-datatables will only support datatables v2 starting from version X.

I was only expecting a small quote on the FAQ page. What you implemented is way better :+1: 

However, I think we need to invert the sources, i.e. `md(HTML|TS)` for source on DT v2 and have a `md(HTML|TS)DTV1` for V1 instead.

Indeed, we will update and maintain the documentation for DT v2 and the ones for DT V1 will not be touched (or will be deleted in the future). In other words, we should use variable names with suffix/prefix for special cases and have the "standard" variable name for the documentation we need to keep up-to-date.

The benefit is double:

- it's implicit that "standard" variable name like "mdHTML" is the main one
- if we delete the documentation for legacy DT v1, it's easier to do it by searching the keyword "v1" (or legacy?)

What do you think?

File: demo/src/app/extensions/responsive-extension.component.html (hunk 8 above)
Comment (shanmukhateja) [reply demo/src/app/extensions/responsive-extension.component.html]: Fixed :)

**Comment by shanmukhateja** [comment]
> > **chore**: We need another change on the documentation where we need to specify that angular-datatables will only support datatables v2 starting from version X.
> 
> I was only expecting a small quote on the FAQ page. What you implemented is way better 👍
> 
> However, I think we need to invert the sources, i.e. `md(HTML|TS)` for source on DT v2 and have a `md(HTML|TS)DTV1` for V1 instead.
> 
> Indeed, we will update and maintain the documentation for DT v2 and the ones for DT V1 will not be touched (or will be deleted in the future). In other words, we should use variable names with suffix/prefix for special cases and have the "standard" variable name for the documentation we need to keep up-to-date.
> 
> The benefit is double:
> 
>     * it's implicit that "standard" variable name like "mdHTML" is the main one
> 
>     * if we delete the documentation for legacy DT v1, it's easier to do it by searching the keyword "v1" (or legacy?)
> 
> 
> What do you think?

In hindsight, I should've done it this way.

Fixed in [`e473f9d` (#1783)](https://github.com/l-lin/angular-datatables/pull/1783/commits/e473f9d9a48d1a4740a7182b41d0485059894400)

**Review by l-lin — APPROVED** [review]

**Pull request merged**